nemo/collections/nlp/models/question_answering/qa_s2s_model.py

```diff
@@ -0,0 +1,447 @@
+            "shuffle": False,
+            "num_samples": num_samples,
+            'num_workers': num_workers,
+            'pin_memory': False,
+            'drop_last': False,
+        }
+        dataloader_cfg = OmegaConf.create(dataloader_cfg)
+        inference_dl = self._setup_dataloader_from_config(cfg=dataloader_cfg, mode=INFERENCE_MODE)
+
+        return inference_dl
+
+    def training_step(self, batch, batch_idx):
+        input_ids, input_attn_mask, unique_ids, labels = batch
+        loss, _ = self.forward(input_ids, input_attn_mask, labels)
+
+        self.log("train_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        return {'loss': loss}
+
+    def validation_step(self, batch, batch_idx):
+        if self.trainer.testing:
+            prefix = 'test'
+        else:
+            prefix = 'val'
+
+        input_ids, input_attn_mask, unique_ids, labels = batch
+        loss, per_sample_perplexity = self.forward(input_ids, input_attn_mask, labels)
+        generated_answers = self._generate_candidates(input_ids, input_attn_mask)
+
+        labels[labels == -100] = self.tokenizer.tokenizer.pad_token_id
+
+        return {
+            "unique_ids": unique_ids,
+            f"{prefix}_loss": loss,
+            "per_sample_perplexity": per_sample_perplexity,
+            "input": self.tokenizer.tokenizer.batch_decode(input_ids, skip_special_tokens=True),
+            "ground_truth_answers": self.tokenizer.tokenizer.batch_decode(labels, skip_special_tokens=True),
+            "generated_answers": generated_answers,
+        }
+
+    def test_step(self, batch, batch_idx):
+        return self.validation_step(batch, batch_idx)
+
+    def validation_epoch_end(self, outputs):
+        if self.trainer.testing:
+            prefix = 'test'
+        else:
+            prefix = 'val'
+
+        loss_terms = [x[f"{prefix}_loss"] for x in outputs]
+        generated_answers, unique_ids, per_sample_perplexity = self._convert_dict_outputs_to_lists(
+            outputs, ["generated_answers", "unique_ids", "per_sample_perplexity"]
+        )
+
+        avg_loss = torch.stack(loss_terms).mean()
+
+        eval_dataset = self._test_dl.dataset if self.trainer.testing else self._validation_dl.dataset
+        exact_match, f1 = self.evaluate(
+            eval_dataset.features, eval_dataset.examples, unique_ids, per_sample_perplexity, generated_answers,
+        )
+
+        logging.info(f"{prefix} exact match {exact_match}")
+        logging.info(f"{prefix} f1 {f1}")
+
+        self.log(f'{prefix}_loss', avg_loss)
+        self.log(f'{prefix}_exact_match', exact_match)
+        self.log(f'{prefix}_f1', f1)
+
+    def test_epoch_end(self, outputs):
+        self.validation_epoch_end(outputs)
+
+    @typecheck()
+    def forward(self, input_ids, input_attn_mask, labels):
+        loss, per_sample_perplexity = None, None
+        if self.cfg.library == "huggingface":
+            with autocast(enabled=False):
```

Comment:
not sure about details of autocast implementation, but it looks to me like you're using the autocast context with enabled=False? If that's the case, maybe we should just remove it?

Reply:
Removing it causes problems when running the model in mixed precision / fp16. I am yet to dig into the details as what this does, but this was pointed out as a bug in HF for T5/BART models along with the fix as above. Removing it currently won't allow operating the model in fp16

Reply:
I see - if it's a bug like that, you might want to write a comment and link the bug if possible. As far as I am aware, I think there are some layers in T5 (only for T5-large and above) that are in fp32 with the rest in bf16 so they might not be cast-able to fp16 easily. BART was ok for me.